File: lluvia/nodes/lluvia/color/HSVA2RGBA_test.cpp
/**
 * \file HSVA2RGBA_test.cpp
 * \brief Test HSVA2RGBA node.
 * \copyright 2021, Juan David Adarve. See AUTHORS for more details
 * \license Apache 2.0, see LICENSE for more details
 */

#define CATCH_CONFIG_MAIN
#include "catch2/catch.hpp"

#include <cmath>
#include <memory>
#include "lluvia/core.h"
#include "lluvia/cpp/core/_virtual_includes/core_cc_library/lluvia/core.h"

#include "tools/cpp/runfiles/runfiles.h"
using bazel::tools::cpp::runfiles::Runfiles;

TEST_CASE("goodUse", "HSVA2RGBA_test") {

    auto runfiles = Runfiles::CreateForTest(nullptr);
    REQUIRE(runfiles != nullptr);

    ///////////////////////////////////////////////////////
    // Create a session and a memory
    ///////////////////////////////////////////////////////
    auto session = ll::Session::create(ll::SessionDescriptor().enableDebug(true));
    REQUIRE(session != nullptr);

    auto memory = session->createMemory(ll::MemoryPropertyFlagBits::DeviceLocal, 0, false);
    REQUIRE(memory != nullptr);

    ///////////////////////////////////////////////////////
    // Register program and builder
    ///////////////////////////////////////////////////////
    auto program = session->createProgram(runfiles->Rlocation("lluvia/lluvia/nodes/lluvia/color/HSVA2RGBA.spv"));
    REQUIRE(program != nullptr);
    session->setProgram("lluvia/color/HSVA2RGBA", program);

    REQUIRE_NOTHROW(session->scriptFile(runfiles->Rlocation("lluvia/lluvia/nodes/lluvia/color/HSVA2RGBA.lua")));

    ///////////////////////////////////////////////////////
    // Create the inputs
    ///////////////////////////////////////////////////////
    auto in_hsva_imgDesc = ll::ImageDescriptor{}
            .setWidth(640)
            .setHeight(480)
            .setChannelType(ll::ChannelType::Float32)
            .setChannelCount(ll::ChannelCount::C4)
            .setUsageFlags(vk::ImageUsageFlagBits::eStorage | vk::ImageUsageFlagBits::eSampled | vk::ImageUsageFlagBits::eTransferDst);

    auto in_hsva_viewDesc = ll::ImageViewDescriptor {}
            .setFilterMode(ll::ImageFilterMode::Nearest)
            .setAddressMode(ll::ImageAddressMode::ClampToEdge)
            .setNormalizedCoordinates(false);

    auto in_hsva = ll::createAndInitImageView(session, memory, in_hsva_imgDesc, in_hsva_viewDesc);
    REQUIRE(in_hsva != nullptr);

    ///////////////////////////////////////////////////////
    // Create the node
    ///////////////////////////////////////////////////////
    auto node = std::shared_ptr<ll::ComputeNode> {nullptr};

    REQUIRE_NOTHROW(([&]{
        node = session->createComputeNode("lluvia/color/HSVA2RGBA");
        })());

    REQUIRE_NOTHROW(node->bind("in_hsva", in_hsva));
    REQUIRE_NOTHROW(node->init());

    ///////////////////////////////////////////////////////
    // Validate out_rgba
    ///////////////////////////////////////////////////////
    auto out_rgba_object = std::shared_ptr<ll::Object> {nullptr};
    REQUIRE_NOTHROW(([&]{
        out_rgba_object = node->getPort("out_rgba");
        })());

    REQUIRE(out_rgba_object != nullptr);
    REQUIRE(out_rgba_object->getType() == ll::ObjectType::ImageView);

    auto out_rgba = std::static_pointer_cast<ll::ImageView>(out_rgba_object);
    REQUIRE(out_rgba->getWidth() == in_hsva->getWidth());
    REQUIRE(out_rgba->getHeight() == in_hsva->getHeight());
    REQUIRE(out_rgba->getDepth() == in_hsva->getDepth());

    REQUIRE(out_rgba->getChannelType() == ll::ChannelType::Uint8);
    REQUIRE(out_rgba->getChannelCount() == ll::ChannelCount::C4);

    ///////////////////////////////////////////////////////
    // Run the node
    ///////////////////////////////////////////////////////
    session->run(*node);

    ///////////////////////////////////////////////////////
    // Validate good usage of Vulkan :)
    ///////////////////////////////////////////////////////
    REQUIRE_FALSE(ll::hasReceivedVulkanWarningMessages());
}

File: lluvia/nodes/lluvia/color/HSVA2RGBA_test.py
import pytest
import numpy as np

import lluvia as ll
import lluvia.util as ll_util
import lluvia_test as ll_test


def test_goodUse():

    nodeName = 'lluvia/color/HSVA2RGBA'

    session = ll.createSession(enableDebug=True, loadNodeLibrary=False)
    ll_test.loadNode(session,
                     builderPath='lluvia/lluvia/nodes/lluvia/color/HSVA2RGBA.lua',
                     programPath='lluvia/lluvia/nodes/lluvia/color/HSVA2RGBA.comp.spv',
                     programName='lluvia/color/HSVA2RGBA.comp'
                     )

    node = session.createComputeNode(nodeName)

    memory = session.createMemory(
        flags=[ll.MemoryPropertyFlagBits.DeviceLocal], pageSize=0)

    imgHSVA = np.zeros((640, 480, 4), dtype=np.float32)
    in_hsva = memory.createImageViewFromHost(imgHSVA)

    node.bind('in_hsva', in_hsva)
    node.init()

    out_rgba = node.getPort('out_rgba')
    assert(out_rgba is not None)
    assert(out_rgba.width == in_hsva.width)
    assert(out_rgba.height == in_hsva.height)
    assert(out_rgba.depth == in_hsva.depth)
    assert(out_rgba.channelType == ll.ChannelType.Uint8)
    assert(out_rgba.channels == 4)

    session.run(node)

    assert(not session.hasReceivedVulkanWarningMessages())


if __name__ == "__main__":

    raise SystemExit(pytest.main([__file__]))
